File: src/editor.ts
#!/usr/bin/env bun
import crypto from "node:crypto";
import fs from "node:fs";
import readline from "node:readline";
import { sanitizeTerminalText, wrapText } from "./format";
import { stateDir } from "./paths";
import { layoutComment } from "./layout";
import { lineEnd, lineStart, resolveEditKey, wordEnd, wordStart } from "./edit-keys";
import { charWidth, stringWidth, truncateToWidth } from "./width";
import type { StoreResult } from "./store";
import {
  parsePendingAnnotation,
  pendingAnnotationFromInvocation,
  type Annotation,
  type PendingAnnotation,
} from "./types";

const pendingPath = process.env.HERDR_ANNOTATE_PENDING;
let pending: PendingAnnotation;

function invocationContext(): unknown {
  try {
    return JSON.parse(process.env.HERDR_PLUGIN_CONTEXT_JSON ?? "{}");
  } catch {
    return {};
  }
}

try {
  const invocation = invocationContext();
  if (!pendingPath) {
    const fallback = pendingAnnotationFromInvocation(invocation);
    if (!fallback) throw new Error("Missing pending annotation");
    pending = fallback;
  } else {
    const decoded: unknown = JSON.parse(fs.readFileSync(pendingPath, "utf8"));
    const parsed = parsePendingAnnotation(decoded);
    if (!parsed) throw new Error("Pending annotation is invalid");
    pending = parsed;
    fs.rmSync(pendingPath, { force: true });
  }
} catch (error) {
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
}

const out = (value: string) => process.stdout.write(value);
const comment: string[] = [];
let cursor = 0;
let status = "";
let finished = false;

function moveCursorVertical(delta: number): void {
  const before = comment.slice(0, cursor).join("");
  const lines = before.split("\n");
  const row = lines.length - 1;
  const col = stringWidth(lines[row] ?? "");
  const allLines = comment.join("").split("\n");
  const targetRow = Math.max(0, Math.min(allLines.length - 1, row + delta));
  let next = 0;
  for (let index = 0; index < targetRow; index += 1) next += Array.from(allLines[index] as string).length + 1;
  // Land on the character whose cell column is closest to the current one.
  let offset = 0;
  let used = 0;
  for (const char of allLines[targetRow] as string) {
    const width = charWidth(char);
    if (used + width > col) break;
    used += width;
    offset += 1;
  }
  cursor = next + offset;
}

function writeAt(row: number, col: number, text: string): void {
  out(`\x1b[${row};${col}H${text}`);
}

function render(): void {
  const cols = Math.max(20, process.stdout.columns || 86);
  const rows = Math.max(10, process.stdout.rows || 22);
  const left = 3;
  const innerWidth = Math.max(1, cols - 4);
  const selectionRows = Math.max(3, Math.min(7, Math.floor((rows - 6) / 2)));
  const editorRows = Math.max(1, rows - selectionRows - 5);
  const wrappedSelection = wrapText(sanitizeTerminalText(pending.selectedText), innerWidth);
  const selected = wrappedSelection.slice(0, selectionRows);
  const editing = layoutComment(comment, cursor, innerWidth);
  const editorStart = Math.max(0, editing.cursorRow - editorRows + 1);
  const visibleEditor = editing.lines.slice(editorStart, editorStart + editorRows);

  out("\x1b[2J\x1b[H\x1b[?25l");
  writeAt(2, left, "\x1b[1mSelected text\x1b[0m");
  selected.forEach((line, index) => writeAt(3 + index, left, `\x1b[2m${line}\x1b[0m`));
  if (wrappedSelection.length > selectionRows) {
    writeAt(2 + selectionRows, left + innerWidth - 1, "\x1b[2m…\x1b[0m");
  }

  const commentTitleRow = 3 + selectionRows;
  writeAt(commentTitleRow, left, "\x1b[1mComment\x1b[0m");
  visibleEditor.forEach((line, index) => writeAt(commentTitleRow + 1 + index, left, line));

  const footer = status || "Ctrl+S save  ·  Esc cancel  ·  Enter new line";
  writeAt(rows, left, `\x1b[2m${truncateToWidth(footer, innerWidth)}\x1b[0m`);

  const visualCursorRow = editing.cursorRow - editorStart;
  if (visualCursorRow >= 0 && visualCursorRow < editorRows) {
    writeAt(commentTitleRow + 1 + visualCursorRow, left + editing.cursorCol, "\x1b[?25h");
  }
}

function cleanup(): void {
  if (finished) return;
  finished = true;
  if (process.stdin.isTTY) process.stdin.setRawMode(false);
  out("\x1b[?25h\x1b[2J\x1b[H\x1b[?1049l");
}

function exit(code: number): void {
  cleanup();
  process.exit(code);
}

async function save(): Promise<void> {
  const value = comment.join("").trim();
  if (!value) {
    status = "Write a comment before saving.";
    render();
    return;
  }
  const dir = stateDir();
  if (!dir) {
    status = "Plugin state directory is unavailable.";
    render();
    return;
  }
  const annotation: Annotation = {
    ...pending,
    id: crypto.randomUUID(),
    comment: value,
    createdAt: new Date().toISOString(),
  };
  let saved: StoreResult<undefined>;
  try {
    const { appendAnnotation } = await import("./store");
    saved = appendAnnotation(dir, annotation);
  } catch {
    status = "Unable to save annotation.";
    render();
    return;
  }
  if (!saved.ok) {
    status = saved.message;
    render();
    return;
  }
  status = "Saved.";
  render();
  setTimeout(() => exit(0), 250);
}

process.on("exit", cleanup);
process.on("SIGTERM", () => exit(0));
try {
  process.on("SIGHUP", () => exit(0));
} catch {
  // SIGHUP is not available on every supported platform.
}
process.stdout.on("resize", render);

readline.emitKeypressEvents(process.stdin, { escapeCodeTimeout: 20 } as any);
if (process.stdin.isTTY) process.stdin.setRawMode(true);
process.stdin.resume();
process.stdin.on("keypress", (text: string, key: readline.Key) => {
  status = "";
  if (key.ctrl && key.name === "c") return exit(0);
  if (key.ctrl && key.name === "s") {
    void save();
    return;
  }
  if (key.name === "escape") return exit(0);
  const action = resolveEditKey(key);
  if (action === "word-left") {
    cursor = wordStart(comment, cursor);
  } else if (action === "word-right") {
    cursor = wordEnd(comment, cursor);
  } else if (action === "line-start") {
    cursor = lineStart(comment, cursor);
  } else if (action === "line-end") {
    cursor = lineEnd(comment, cursor);
  } else if (action === "delete-word") {
    const start = wordStart(comment, cursor);
    comment.splice(start, cursor - start);
    cursor = start;
  } else if (action === "delete-line") {
    const start = lineStart(comment, cursor);
    comment.splice(start, cursor - start);
    cursor = start;
  } else if (key.name === "backspace") {
    if (cursor > 0) comment.splice(--cursor, 1);
  } else if (key.name === "delete") {
    if (cursor < comment.length) comment.splice(cursor, 1);
  } else if (key.name === "left") {
    cursor = Math.max(0, cursor - 1);
  } else if (key.name === "right") {
    cursor = Math.min(comment.length, cursor + 1);
  } else if (key.name === "up") {
    moveCursorVertical(-1);
  } else if (key.name === "down") {
    moveCursorVertical(1);
  } else if (key.name === "home") {
    while (cursor > 0 && comment[cursor - 1] !== "\n") cursor -= 1;
  } else if (key.name === "end") {
    while (cursor < comment.length && comment[cursor] !== "\n") cursor += 1;
  } else if (key.name === "return") {
    comment.splice(cursor, 0, "\n");
    cursor += 1;
  } else if (text && !key.ctrl && !key.meta) {
    const inserted = Array.from(text);
    comment.splice(cursor, 0, ...inserted);
    cursor += inserted.length;
  }
  render();
});

out("\x1b[?1049h");
render();

File: rust/src/editor.rs
//! Interactive comment editor pane.

use std::path::Path;
use std::time::Duration;

use chrono::{SecondsFormat, Utc};
use ratatui::Frame;
use ratatui::crossterm::event::{self, Event, KeyCode, KeyEvent, KeyEventKind, KeyModifiers};
use ratatui::layout::{Position, Rect};
use ratatui::style::{Modifier, Style};
use ratatui::text::Line;
use ratatui::widgets::{Clear, Paragraph};
use serde_json::Value;
use uuid::Uuid;

use crate::edit_keys::{EditAction, line_end, line_start, resolve_edit_key, word_end, word_start};
use crate::format::{sanitize_terminal_text, wrap_text};
use crate::layout::layout_comment;
use crate::paths::state_dir;
use crate::store::{append_annotation, append_annotation_context_first};
use crate::termination::Termination;
use crate::types::{
    Annotation, PendingAnnotation, javascript_trim, parse_pending_annotation,
    pending_annotation_from_invocation,
};
use crate::width::{char_width, string_width, truncate_to_width};

#[cfg(test)]
const DEFAULT_COLS: u16 = 86;
#[cfg(test)]
const DEFAULT_ROWS: u16 = 22;

/// Editor state, independent of the terminal backend.
#[derive(Debug)]
pub struct EditorApp {
    pending: PendingAnnotation,
    saved_field_order: SavedFieldOrder,
    comment: Vec<char>,
    cursor: usize,
    status: String,
    quit: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum SavedFieldOrder {
    CapturedAtThenContext,
    ContextThenCapturedAt,
}

impl EditorApp {
    /// Start an empty comment for a captured selection.
    pub fn new(pending: PendingAnnotation) -> Self {
        Self::with_field_order(pending, SavedFieldOrder::CapturedAtThenContext)
    }

    fn with_field_order(pending: PendingAnnotation, saved_field_order: SavedFieldOrder) -> Self {
        Self {
            pending,
            saved_field_order,
            comment: Vec::new(),
            cursor: 0,
            status: String::new(),
            quit: false,
        }
    }

    /// Draw the same selected-text, comment, and footer regions as the TypeScript editor.
    pub fn draw(&self, frame: &mut Frame<'_>) {
        let area = frame.area();
        frame.render_widget(Clear, area);
        let cols = usize::from(area.width.max(20));
        let rows = usize::from(area.height.max(10));
        let left = 2usize;
        let inner_width = cols.saturating_sub(4).max(1);
        let selection_rows = ((rows.saturating_sub(6)) / 2).clamp(3, 7);
        let editor_rows = rows.saturating_sub(selection_rows + 5).max(1);
        let wrapped_selection = wrap_text(
            &sanitize_terminal_text(&self.pending.selected_text),
            inner_width,
        );
        let selected = wrapped_selection.iter().take(selection_rows);
        let editing = layout_comment(&self.comment, self.cursor, inner_width);
        let editor_start = editing
            .cursor_row
            .saturating_sub(editor_rows.saturating_sub(1));

        render_line(
            frame,
            left,
            1,
            "Selected text",
            inner_width,
            Style::default().add_modifier(Modifier::BOLD),
        );
        for (index, line) in selected.enumerate() {
            render_line(
                frame,
                left,
                2 + index,
                line,
                inner_width,
                Style::default().add_modifier(Modifier::DIM),
            );
        }
        if wrapped_selection.len() > selection_rows {
            render_line(
                frame,
                left + inner_width.saturating_sub(1),
                1 + selection_rows,
                "…",
                1,
                Style::default().add_modifier(Modifier::DIM),
            );
        }

        let comment_title_row = 2 + selection_rows;
        render_line(
            frame,
            left,
            comment_title_row,
            "Comment",
            inner_width,
            Style::default().add_modifier(Modifier::BOLD),
        );
        for (index, line) in editing
            .lines
            .iter()
            .skip(editor_start)
            .take(editor_rows)
            .enumerate()
        {
            render_line(
                frame,
                left,
                comment_title_row + 1 + index,
                line,
                inner_width,
                Style::default(),
            );
        }
        let footer = if self.status.is_empty() {
            "Ctrl+S save  ·  Esc cancel  ·  Enter new line"
        } else {
            &self.status
        };
        render_line(
            frame,
            left,
            rows.saturating_sub(1),
            &truncate_to_width(footer, inner_width),
            inner_width,
            Style::default().add_modifier(Modifier::DIM),
        );

        let visual_row = editing.cursor_row.saturating_sub(editor_start);
        if editing.cursor_row >= editor_start && visual_row < editor_rows {
            let x = left.saturating_add(editing.cursor_col);
            let y = comment_title_row + 1 + visual_row;
            if let (Ok(x), Ok(y)) = (u16::try_from(x), u16::try_from(y))
                && x < area.width
                && y < area.height
            {
                frame.set_cursor_position(Position::new(x, y));
            }
        }
    }

    /// Handle one keyboard event. Returns `true` when a save should be attempted.
    pub fn handle_key(&mut self, key: KeyEvent) -> bool {
        if key.kind == KeyEventKind::Release {
            return false;
        }
        self.status.clear();
        if key.code == KeyCode::Char('c') && key.modifiers.contains(KeyModifiers::CONTROL) {
            self.quit = true;
            return false;
        }
        if key.code == KeyCode::Char('s') && key.modifiers.contains(KeyModifiers::CONTROL) {
            return true;
        }
        if let Some(action) = resolve_edit_key(&key) {
            self.apply_edit_action(action);
            return false;
        }
        match key.code {
            KeyCode::Esc => self.quit = true,
            KeyCode::Backspace => {
                if self.cursor > 0 {
                    self.cursor -= 1;
                    self.comment.remove(self.cursor);
                }
            }
            KeyCode::Delete => {
                if self.cursor < self.comment.len() {
                    self.comment.remove(self.cursor);
                }
            }
            KeyCode::Left => self.cursor = self.cursor.saturating_sub(1),
            KeyCode::Right => self.cursor = (self.cursor + 1).min(self.comment.len()),
            KeyCode::Up => self.move_cursor_vertical(-1),
            KeyCode::Down => self.move_cursor_vertical(1),
            KeyCode::Home => {
                while self.cursor > 0 && self.comment.get(self.cursor - 1) != Some(&'\n') {
                    self.cursor -= 1;
                }
            }
            KeyCode::End => {
                while self.cursor < self.comment.len()
                    && self.comment.get(self.cursor) != Some(&'\n')
                {
                    self.cursor += 1;
                }
            }
            KeyCode::Enter => self.insert('\n'),
            KeyCode::Char(character)
                if !key
                    .modifiers
                    .intersects(KeyModifiers::CONTROL | KeyModifiers::ALT) =>
            {
                self.insert(character);
            }
            _ => {}
        }
        false
    }

    /// Apply a word or line action, mirroring the retired Bun editor's `resolveEditKey` branches.
    fn apply_edit_action(&mut self, action: EditAction) {
        match action {
            EditAction::WordLeft => self.cursor = word_start(&self.comment, self.cursor),
            EditAction::WordRight => self.cursor = word_end(&self.comment, self.cursor),
            EditAction::LineStart => self.cursor = line_start(&self.comment, self.cursor),
            EditAction::LineEnd => self.cursor = line_end(&self.comment, self.cursor),
            EditAction::DeleteWord => {
                let start = word_start(&self.comment, self.cursor);
                self.comment.drain(start..self.cursor);
                self.cursor = start;
            }
            EditAction::DeleteLine => {
                let start = line_start(&self.comment, self.cursor);
                self.comment.drain(start..self.cursor);
                self.cursor = start;
            }
        }
    }

    fn insert(&mut self, character: char) {
        self.comment.insert(self.cursor, character);
        self.cursor += 1;
    }

    fn move_cursor_vertical(&mut self, delta: isize) {
        let before = self.comment.iter().take(self.cursor).collect::<String>();
        let row = before.split('\n').count().saturating_sub(1);
        let col = before.rsplit('\n').next().map_or(0, string_width);
        let joined = self.comment.iter().collect::<String>();
        let lines = joined.split('\n').collect::<Vec<_>>();
        let target_row = row
            .saturating_add_signed(delta)
            .min(lines.len().saturating_sub(1));
        let mut next = lines
            .iter()
            .take(target_row)
            .map(|line| line.chars().count() + 1)
            .sum::<usize>();
        let mut used = 0;
        for character in lines.get(target_row).copied().unwrap_or_default().chars() {
            let width = char_width(character);
            if used + width > col {
                break;
            }
            used += width;
            next += 1;
        }
        self.cursor = next;
    }

    fn save(&mut self, dir: Option<&Path>) -> bool {
        let value = self.comment.iter().collect::<String>();
        let value = javascript_trim(&value).to_owned();
        if value.is_empty() {
            "Write a comment before saving.".clone_into(&mut self.status);
            return false;
        }
        let Some(dir) = dir else {
            "Plugin state directory is unavailable.".clone_into(&mut self.status);
            return false;
        };
        let annotation = Annotation::from_pending(
            self.pending.clone(),
            Uuid::new_v4().to_string(),
            value,
            now_iso(),
        );
        let saved = match self.saved_field_order {
            SavedFieldOrder::CapturedAtThenContext => append_annotation(dir, &annotation),
            SavedFieldOrder::ContextThenCapturedAt => {
                append_annotation_context_first(dir, &annotation)
            }
        };
        match saved {
            Ok(()) => {
                "Saved.".clone_into(&mut self.status);
                true
            }
            Err(message) => {
                self.status = message;
                false
            }
        }
    }
}

fn render_line(frame: &mut Frame<'_>, x: usize, y: usize, text: &str, width: usize, style: Style) {
    let (Ok(x), Ok(y), Ok(width)) = (u16::try_from(x), u16::try_from(y), u16::try_from(width))
    else {
        return;
    };
    let area = frame.area();
    if x >= area.width || y >= area.height {
        return;
    }
    let width = width.min(area.width.saturating_sub(x));
    frame.render_widget(
        Paragraph::new(Line::styled(text.to_owned(), style)),
        Rect::new(x, y, width, 1),
    );
}

fn now_iso() -> String {
    Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true)
}

pub(crate) fn now_iso_for_manager() -> String {
    now_iso()
}

fn invocation_context() -> Value {
    std::env::var("HERDR_PLUGIN_CONTEXT_JSON")
        .ok()
        .and_then(|text| serde_json::from_str(&text).ok())
        .unwrap_or_else(|| Value::Object(serde_json::Map::new()))
}

fn pending_from_env() -> Result<(PendingAnnotation, SavedFieldOrder), String> {
    let invocation = invocation_context();
    let Some(path) = std::env::var_os("HERDR_ANNOTATE_PENDING").filter(|value| !value.is_empty())
    else {
        return pending_annotation_from_invocation(&invocation, now_iso())
            .map(|pending| (pending, SavedFieldOrder::ContextThenCapturedAt))
            .ok_or_else(|| "Missing pending annotation".to_owned());
    };
    let path = std::path::PathBuf::from(path);
    let text = std::fs::read_to_string(&path).map_err(|error| error.to_string())?;
    let decoded = serde_json::from_str::<Value>(&text).map_err(|error| error.to_string())?;
    let pending = parse_pending_annotation(&decoded)
        .ok_or_else(|| "Pending annotation is invalid".to_owned())?;
    remove_pending_file(&path)?;
    Ok((pending, SavedFieldOrder::CapturedAtThenContext))
}

fn remove_pending_file(path: &Path) -> Result<(), String> {
    match std::fs::remove_file(path) {
        Ok(()) => Ok(()),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(error) => Err(error.to_string()),
    }
}

/// Run the interactive editor pane from Herdr's environment.
pub fn run() -> Result<(), String> {
    let (pending, saved_field_order) = pending_from_env()?;
    let mut app = EditorApp::with_field_order(pending, saved_field_order);
    let dir = state_dir();
    let termination = Termination::install();
    let mut terminal = ratatui::init();
    let result = (|| -> Result<(), String> {
        while !app.quit && !termination.requested() {
            terminal
                .draw(|frame| app.draw(frame))
                .map_err(|error| error.to_string())?;
            while !termination.requested()
                && !event::poll(Duration::from_millis(50)).map_err(|error| error.to_string())?
            {
            }
            if termination.requested() {
                break;
            }
            let event = event::read().map_err(|error| error.to_string())?;
            if let Event::Key(key) = event
                && app.handle_key(key)
                && app.save(dir.as_deref())
            {
                terminal
                    .draw(|frame| app.draw(frame))
                    .map_err(|error| error.to_string())?;
                std::thread::sleep(Duration::from_millis(250));
                app.quit = true;
            }
        }
        Ok(())
    })();
    ratatui::restore();
    result
}

#[cfg(test)]
mod tests {
    #![allow(clippy::expect_used, reason = "tests assert by panicking")]

    use std::sync::atomic::{AtomicUsize, Ordering};

    use ratatui::Terminal;
    use ratatui::backend::TestBackend;

    use crate::types::InvocationContext;

    use super::*;

    static NEXT_DIR: AtomicUsize = AtomicUsize::new(0);

    fn app() -> EditorApp {
        EditorApp::new(PendingAnnotation {
            selected_text: "first selected line\nsecond line".to_owned(),
            context: InvocationContext::default(),
            captured_at: "captured".to_owned(),
        })
    }

    fn draw(app: &EditorApp) -> Vec<String> {
        let mut terminal =
            Terminal::new(TestBackend::new(DEFAULT_COLS, DEFAULT_ROWS)).expect("terminal");
        terminal.draw(|frame| app.draw(frame)).expect("draw");
        let buffer = terminal.backend().buffer();
        (0..buffer.area.height)
            .map(|y| {
                (0..buffer.area.width)
                    .filter_map(|x| buffer.cell((x, y)))
                    .map(|cell| cell.symbol().to_owned())
                    .collect()
            })
            .collect()
    }

    #[test]
    fn headless_frame_contains_selection_editor_and_keys() {
        let rows = draw(&app());
        assert!(rows.iter().any(|row| row.contains("Selected text")));
        assert!(rows.iter().any(|row| row.contains("first selected line")));
        assert!(rows.iter().any(|row| row.contains("Comment")));
        assert!(rows.iter().any(|row| row.contains("Ctrl+S save")));
    }

    #[test]
    fn editing_keys_and_empty_save_match_the_typescript_editor() {
        let mut editor = app();
        assert!(!editor.handle_key(KeyEvent::from(KeyCode::Char('한'))));
        assert!(!editor.handle_key(KeyEvent::from(KeyCode::Char('a'))));
        assert_eq!(editor.comment, ['한', 'a']);
        editor.handle_key(KeyEvent::from(KeyCode::Left));
        editor.handle_key(KeyEvent::from(KeyCode::Backspace));
        assert_eq!(editor.comment, ['a']);
        let mut empty = app();
        assert!(!empty.save(None));
        assert_eq!(empty.status, "Write a comment before saving.");
    }

    #[test]
    fn word_and_line_keys_match_the_typescript_editor() {
        let mut editor = app();
        for character in "alpha beta".chars() {
            editor.handle_key(KeyEvent::from(KeyCode::Char(character)));
        }
        editor.handle_key(KeyEvent::new(KeyCode::Left, KeyModifiers::ALT));
        assert_eq!(editor.cursor, 6);
        editor.handle_key(KeyEvent::new(KeyCode::Char('b'), KeyModifiers::ALT));
        assert_eq!(editor.cursor, 0);
        editor.handle_key(KeyEvent::new(KeyCode::Char('f'), KeyModifiers::ALT));
        assert_eq!(editor.cursor, 5);
        editor.handle_key(KeyEvent::new(KeyCode::Right, KeyModifiers::CONTROL));
        assert_eq!(editor.cursor, 10);
        editor.handle_key(KeyEvent::new(KeyCode::Left, KeyModifiers::SUPER));
        assert_eq!(editor.cursor, 0);
        editor.handle_key(KeyEvent::new(KeyCode::Right, KeyModifiers::SUPER));
        assert_eq!(editor.cursor, 10);
        editor.handle_key(KeyEvent::new(KeyCode::Char('a'), KeyModifiers::CONTROL));
        assert_eq!(editor.cursor, 0);
        editor.handle_key(KeyEvent::new(KeyCode::Char('e'), KeyModifiers::CONTROL));
        assert_eq!(editor.cursor, 10);
        editor.handle_key(KeyEvent::new(KeyCode::Char('w'), KeyModifiers::CONTROL));
        assert_eq!(editor.comment.iter().collect::<String>(), "alpha ");
        assert_eq!(editor.cursor, 6);
        editor.handle_key(KeyEvent::new(KeyCode::Backspace, KeyModifiers::ALT));
        assert_eq!(editor.comment.iter().collect::<String>(), "");
        assert_eq!(editor.cursor, 0);
    }

    #[test]
    fn delete_line_kills_only_the_current_line() {
        let mut editor = app();
        for character in "one two".chars() {
            editor.handle_key(KeyEvent::from(KeyCode::Char(character)));
        }
        editor.handle_key(KeyEvent::from(KeyCode::Enter));
        for character in "한글 three".chars() {
            editor.handle_key(KeyEvent::from(KeyCode::Char(character)));
        }
        editor.handle_key(KeyEvent::new(KeyCode::Char('u'), KeyModifiers::CONTROL));
        assert_eq!(editor.comment.iter().collect::<String>(), "one two\n");
        assert_eq!(editor.cursor, 8);
        editor.handle_key(KeyEvent::new(KeyCode::Left, KeyModifiers::ALT));
        assert_eq!(editor.cursor, 7);
        editor.handle_key(KeyEvent::new(KeyCode::Left, KeyModifiers::ALT));
        assert_eq!(editor.cursor, 4);
        editor.handle_key(KeyEvent::new(KeyCode::Right, KeyModifiers::ALT));
        assert_eq!(editor.cursor, 7);
        editor.handle_key(KeyEvent::new(KeyCode::Right, KeyModifiers::ALT));
        assert_eq!(editor.cursor, 8);
    }

    #[test]
    fn escape_and_control_c_quit() {
        let mut escape = app();
        escape.handle_key(KeyEvent::from(KeyCode::Esc));
        assert!(escape.quit);
        let mut control = app();
        control.handle_key(KeyEvent::new(KeyCode::Char('c'), KeyModifiers::CONTROL));
        assert!(control.quit);
    }

    #[test]
    fn invocation_fallback_save_uses_its_typescript_property_order() {
        let sequence = NEXT_DIR.fetch_add(1, Ordering::Relaxed);
        let dir = std::env::temp_dir().join(format!(
            "herdr-annotate-editor-{}-{sequence}",
            std::process::id()
        ));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).expect("temporary directory");
        let mut editor = EditorApp::with_field_order(
            PendingAnnotation {
                selected_text: "selection".to_owned(),
                context: InvocationContext::default(),
                captured_at: "captured".to_owned(),
            },
            SavedFieldOrder::ContextThenCapturedAt,
        );
        editor.comment = "comment".chars().collect();
        editor.cursor = editor.comment.len();

        assert!(editor.save(Some(&dir)));
        let saved = std::fs::read_to_string(dir.join("annotations.jsonl")).expect("saved record");
        assert!(saved.starts_with(
            "{\"selectedText\":\"selection\",\"context\":{},\"capturedAt\":\"captured\","
        ));
        let _ = std::fs::remove_dir_all(dir);
    }

    #[test]
    fn pending_removal_is_forceful_like_typescript() {
        let missing = std::env::temp_dir().join(format!(
            "herdr-annotate-missing-pending-{}",
            std::process::id()
        ));
        let _ = std::fs::remove_file(&missing);
        assert_eq!(remove_pending_file(&missing), Ok(()));
    }
}
